### ISSUE TITLE
Jetty 12: Improve BufferedResponseHandler

### ISSUE THREAD

**Issue by sbordet** [eclipse/jetty.project#10476]
**Jetty version(s)**
12.0.1

**Jetty Environment**
core

**Description**
Jetty 12 core does not buffer response writes, which means that in common cases the response incurs in 2 issues:
* 2 system calls instead of 1 for small writes (one for the write of the content with last=false, one with last=true).
* the client has a more complicated parsing to do, affecting the response times.

`BufferedResponseHandler` allows buffering response writes similarly to what we do with the EE implementations in `HttpOutput` but should be revisited.

**Comment by sunng87** [comment]
There is up to 20% throughput gain by buffering accorfing to our tests. It world be nice to have buffering built-in again.

But I got an issue report with Server-Sent Event not working on my Jetty 11 implementation. Not sure if it's related to buffering mechanism. I will be back with more details.

**Comment by gregw** [comment]
@sbordet whilst I'm not disputing the need for a buffering handler sometimes, I don't agree that the common case is for 2 writes.

If there is only a single write to be done, then the solution is to pass that with last==true, not copy it into a buffer and then write it later.

There will also be many handlers that will generate a response into something like a StringBuilder or a writer with internal buffering.   We don't want to end up buffering content that has already been buffered.

I think it would be far better to start with a `BufferingContentSink` that could be applied by a particular handler that might be generating content in multiple byte arrays.     By all means we can then have a utility handler that uses this in a wrapper on all responses, but I think that is secondary.

Similarly we should probably provide more examples that show something like `new BufferedOutputStream(Content.Sink.asOutputStream(response))` or perhaps we can provide a `Content.Sink#asBufferedOutputStream(Content.Sink, int bufferSize)` API.

Perhaps we should also look at popular JSON libraries (and our own) and see if there are efficient ways to send a `JsonObject` with converting it to a string, then to bytes and then perhaps buffering to make efficient writes.

In summary, we should make sure that we provide APIs and examples of how to efficiently generate content before we just offer a "buffer every response" as the primary solution.

**Comment by gregw** [comment]
@sunng87 I've had a little look at https://github.com/sunng87/ring-jetty9-adapter/pull/106 .... but clojure is a bit beyond me at the moment.   I can see where you are defining classes that extend Handler$ABstract, but I can't really work out exactly how you are writing content to the responses.   It may well be that providing a buffering handler is best for your use-case, but perhaps doing it in the generation would be better.  Can you describe how the content is actually generated?

**Comment by sunng87** [comment]
No problem. What my project has being doing is to create a Jetty adapter for Clojure's [Ring spec](https://github.com/ring-clojure/ring/blob/master/SPEC), which you can think it as light-weighted servlet, or WSGI in python. So the content is actually generated from user code. It can be anything implemented Ring's `StreamableResponseBody` protocol. For example, a Ring handler:

```clojure
(defn handler [req]
  {:status 200 :body "<h1>It works</h1>"})
```

By default, Clojure has these types implemented `StreamableResponseBody`: https://github.com/ring-clojure/ring/blob/master/SPEC#L130 , and its implementation shows how these data written into the stream: https://github.com/ring-clojure/ring/blob/master/ring-core/src/ring/core/protocols.clj#L20

We realized that using buffered response as in Jetty 11 has 20% performance gain than Jetty 12's defaults.

**Comment by gregw** [comment]
@sunng87 as @lorban has pointed out to me offline, you might also be suffering from a lack of content-length.   Having a buffering handler can help with length determination, as any responses that fit inside a single buffer may have their content-length determined.

However, in regards to my comments above, I realise that there is a problem if a solution just used something like a `BufferingOutputStream` as close will be implemented as a flush then a close.   We probably need to offer our own version of a buffering output stream that can make a close be implemented with a flush-with-close that our API supports.

**Comment by lorban** [comment]
@sunng87 I noticed we already include a basic buffering handler in Jetty 12.0.1: `org.eclipse.jetty.server.handler.BufferedResponseHandler`. You can make use of it by wrapping your own `SyncProxyHandler` with it; in java it would look like this:

```java
Handler h = new BufferedResponseHandler(new SyncProxyHandler());
```

Could you please give it a try and report back the outcome? Thanks!

**Comment by sunng87** [comment]
@gregw Yes, our tests saw [performance regression](https://github.com/sunng87/ring-jetty9-adapter/pull/106#issuecomment-1703649545) from Jetty 11 to Jetty 12 with default settings. Because our response body was pretty small (~10 bytes), I assume `content-length` vs chunked encoding can be the key difference. The idea for this issue is to fix the regression.

Perhaps you can test on Jetty 12 to check the difference of `Content.Sink.write(response, true, "it works", callback);` which generates content-length, and `output.write("it works");` which creates chunked encoding. 

By "buffering mechanism" in this thread, I assume it's the mechanism to generate `content-length` as `HttpOutput` in Jetty 11.

**Comment by sunng87** [comment]
@lorban cool, I will try it later today.

**Comment by sunng87** [comment]
@lorban I confirm `BufferedResponseHandler` enables content-length and improves benchmark quite a bit. Now our clojure app on Jetty 12 is about 10% faster than Jetty 11.

I think we can close this issue now. Thank you for your help @gregw @lorban @sbordet

**Comment by gregw** [comment]
@sunng87 awesome. 

But let's leave this open for a bit as I think the is more we can do to ensure appropriate buffering. I also think we might need to do more to address the double write on close in some instances.

**Comment by lorban** [comment]
The existing `BufferedResponseHandler` contains the basic logic to buffer responses, but it clearly needs some improvements, like for instance, the fact that it buffers up to 2 GB per request with no straightforward way to configure that. Plus, it introduces an extra memory copy that we could get rid of in many cases.

Let's keep this issue open but reword its title to say that it's about improving the existing handler.

**Comment by sbordet** [comment]
Fixed by #10483.

**Comment by lorban** [comment]
@sunng87 we finished the buffered response improvements. On top of the existing `BufferedResponseHandler`, which works exactly like it used to, we also added a few more improvements you might want to make use of:

 - The `BufferedResponseHandler`'s buffer is now controlled by `HttpConfiguration.getOutputBufferSize()` and `HttpConfiguration.getOutputAggregationSize()` so you don't need to set the `BUFFER_SIZE_ATTRIBUTE_NAME` attribute anymore to control it.
 - We added the `Response.asBufferedSink()` and `Response.asBufferedOutputStream()` static helpers so you have the choice between using `BufferedResponseHandler` or directly using a buffered sink or output stream. As for `BufferedResponseHandler`, the buffer is controlled by the same two settings of `HttpConfiguration`.

Feel free to make use of these new features or to keep your implementation the way it was, both should work exactly the same.

**Comment by sunng87** [comment]
Thank you @lorban. I will try out these new APIs with 12.0.2 release.
